src/complex-list-field/complex-list-field.component.ts

```diff
@@ -42,30 +43,122 @@ import { AppGlobalsService, JsonStoreService } from '../shared/services';
+    public jsonStoreService: JsonStoreService,
+    public domUtilService: DomUtilService) {
     super(appGlobalsService, jsonStoreService);
   }
 
+  ngOnInit() {
+    this.navigator = this.schema['x_editor_long_list_navigator'];
+    if (this.navigator) {
+      this.paginatedIndices = this.getIndicesForPage(this.currentPage);
+    } else {
+      // set all indices as paginated indices if pagination is not enabled.
+      this.paginatedIndices = this.values.keySeq().toArray();
+    }
+    this.keys = this.getKeysForCurrentPage();
+  }
+
+  ngOnChanges(changes: SimpleChanges) {
+    let valuesChange = changes['values'];
+    if (valuesChange && !valuesChange.isFirstChange()) {
+      if (!this.navigator && valuesChange.currentValue.size !== valuesChange.previousValue.size) {
+        this.paginatedIndices = this.values.keySeq().toArray();
+      }
+      this.keys = this.getKeysForCurrentPage();
+    }
+  }
+
   onFieldAdd(field: string, index: number) {
     this.keys[index].push(field);
   }
 
-  ngOnChanges(changes: SimpleChanges) {
-    let valuesChanges = changes['values'];
-    if (valuesChanges) {
-      this.keys = valuesChanges.currentValue
-        .map(value => value.keySeq().toArray())
-        .toArray();
+  onFindClick() {
+    // clear for new search
+    this.foundIndices = [];
+    this.currentFound = 0;
+    // search to look for the first match
+    if (this.navigator.findSingle) {
+      let foundIndex = this.values
+        .findIndex(value => this.navigator.findSingle(value, this.findExpression));
+      if (foundIndex > -1) {
+        this.foundIndices.push(foundIndex);
+      }
+    }
+    // search to look for all matches
+    if (this.foundIndices.length === 0 && this.navigator.findMultiple) {
+      this.values
+        .forEach((value, index) => {
+          if (this.navigator.findMultiple(value, this.findExpression)) {
+            this.foundIndices.push(index);
+          }
+        });
+    }
+    // navigate to first search result if found any
+    if (this.foundIndices.length > 0) {
+      this.navigateToItem(this.foundIndices[0]);
+      this.shouldDisplayFoundNavigation = true;
+    } else {
+      this.shouldDisplayFoundNavigation = false;
+    }
+  }
+
+  onFindInputKeypress(key: string) {
+    if (key === 'Enter' ) {
+      this.onFindClick();
+    }
+  }
+
+  onFoundNavigate(direction: number) {
+    // No bound checks, since the buttons are disabled in these cases.
+    this.currentFound += direction;
+    this.navigateToItem(this.foundIndices[this.currentFound]);
+  }
+
+  navigateToItem(index: number) {
+    this.currentPage = Math.floor((index / this.navigator.itemsPerPage) + 1);
+    let itemId = this.path
+      .concat(index)
+      .join('.');
+    setTimeout(() => this.domUtilService.focusAndSelectFirstInputChildById(itemId));
+  }
+
+  onPageChange(page: number) {
+    this.paginatedIndices = this.getIndicesForPage(page);
+    this.keys = this.getKeysForCurrentPage();
+  }
+
```

Comment:
Below here could go to a service but, may be we can do it later, when we need to use these elsewhere.